File: leetcode_python/String/uncommon-words-from-two-sentences.py
"""

884. Uncommon Words from Two Sentences
Easy

A sentence is a string of single-space separated words where each word consists only of lowercase letters.

A word is uncommon if it appears exactly once in one of the sentences, and does not appear in the other sentence.

Given two sentences s1 and s2, return a list of all the uncommon words. You may return the answer in any order.

Example 1:

Input: s1 = "this apple is sweet", s2 = "this apple is sour"
Output: ["sweet","sour"]
Explanation:

The word "sweet" appears only in s1, while the word "sour" appears only in s2.

Example 2:

Input: s1 = "apple apple", s2 = "banana"
Output: ["banana"]

Constraints:

1 <= s1.length, s2.length <= 200
s1 and s2 consist of lowercase English letters and spaces.
s1 and s2 do not have leading or trailing spaces.
All the words in s1 and s2 are separated by a single space.

"""

# V0 

# V1
# https://blog.csdn.net/fuxuemingzhu/article/details/81749170
# time = O(m + n)
# space = O(m + n)
class Solution:
    def uncommonFromSentences(self, A, B):
        """
        :type A: str
        :type B: str
        :rtype: List[str]
        """
        count_A = collections.Counter(A.split(' '))
        count_B = collections.Counter(B.split(' '))
        words = list((count_A.keys() | count_B.keys()) - (count_A.keys() & count_B.keys()))
        ans = []
        for word in words:
            if count_A[word] == 1 or count_B[word] == 1:
                ans.append(word)
        return ans

# V2
# time = O(m + n)
# space = O(m + n)
import collections
class Solution(object):
    def uncommonFromSentences(self, A, B):
        """
        :type A: str
        :type B: str
        :rtype: List[str]
        """
        count = collections.Counter(A.split())
        count += collections.Counter(B.split())
        return [word for word in count if count[word] == 1]

File: leetcode_java/src/main/java/LeetCodeJava/String/UncommonWordsFromTwoSentences.java
package LeetCodeJava.String;

// https://leetcode.com/problems/uncommon-words-from-two-sentences/

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 *  884. Uncommon Words from Two Sentences
 *  Easy
 *
 *  A sentence is a string of single-space separated words where each word
 *  consists only of lowercase letters.
 *
 *  A word is uncommon if it appears exactly once in one of the sentences, and
 *  does not appear in the other sentence.
 *
 *  Given two sentences s1 and s2, return a list of all the uncommon words.
 *  You may return the answer in any order.
 *
 *  Example 1:
 *  Input: s1 = "this apple is sweet", s2 = "this apple is sour"
 *  Output: ["sweet","sour"]
 *
 *  Example 2:
 *  Input: s1 = "apple apple", s2 = "banana"
 *  Output: ["banana"]
 *
 *  Constraints:
 *   - 1 <= s1.length, s2.length <= 200
 */
public class UncommonWordsFromTwoSentences {

    // V0
    // IDEA: HASHMAP COUNT - merge both sentences and keep words with count == 1.
    /**
     * time = O(m + n)
     * space = O(m + n)
     */
    public String[] uncommonFromSentences(String s1, String s2) {
        Map<String, Integer> cnt = new HashMap<>();
        for (String w : s1.split(" ")) {
            cnt.put(w, cnt.getOrDefault(w, 0) + 1);
        }
        for (String w : s2.split(" ")) {
            cnt.put(w, cnt.getOrDefault(w, 0) + 1);
        }

        List<String> res = new ArrayList<>();
        for (Map.Entry<String, Integer> e : cnt.entrySet()) {
            if (e.getValue() == 1) {
                res.add(e.getKey());
            }
        }
        return res.toArray(new String[0]);
    }
}
